Improvements for WeMo Insight switches

deftdawg opened the pull request:
**Description:**
* Changes current power units to watts
* Adds power on times and additional totals

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>
* Partially addresses messed up labels and adds missing Insight data fields #3506
* Recommend patch (https://github.com/pavoni/pywemo/issues/64) to resolve lack of updates for WeMo Insight devices

deftdawg committed:
Improvements for WeMo Insight switches

* Changes current power units to watts
* Adds power on times and additional totals

homeassistant commented:
Hi @jumpkick, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#6001).

Once you do that we will be able to review and accept this pull request.

Thanks!

deftdawg committed:
Update wemo.py

* Reordered datetime import
* Spaces by 4

deftdawg committed:
Update wemo.py

* continuation line under-indented for visual indent

deftdawg committed:
Update wemo.py

trailing whitespace... (argh... the bot should just trim it)

deftdawg committed:
Back to you travis....

deftdawg committed:
Update wemo.py

deftdawg committed:
Update wemo.py

deftdawg committed:
Update wemo.py

balloob commented on homeassistant/components/switch/wemo.py:
```diff
@@ -109,23 +120,88 @@ def device_state_attributes(self):
+    def _current_power_mw(self):
+        """Current power usage in mW."""
+        if self.insight_params:
+            try:
```
Instead of try … catch, use `self.insight_param.get('current_power')`

jumpkick replied:
guess I could do `if self.insight_params and self.insight_param.get('current_power'):`, guessing the original code `if self.insight_params` condition because non-Insight switches don't have those attribs.

pavoni replied:
Exactly right.

balloob commented on homeassistant/components/switch/wemo.py:
```diff
@@ -109,23 +120,88 @@ def device_state_attributes(self):
 
         return attr
 
+#    @property
```
Stale comment ?

jumpkick replied:
Yeah, not needed, was from originally when that method was currrent_power_mwh

balloob commented on homeassistant/components/switch/wemo.py:
```diff
@@ -109,23 +120,88 @@ def device_state_attributes(self):
+            except TypeError:
+                return None
+
+    @property
```
Why are somethings properties and other methods ? Also, why are they properties/method to begin with and not just all part of `device_state_attributes` ?

jumpkick replied:
Started out cutting and pasting what was there (I'm not a python guy, nor do I know HA components)... If we don't need the methods, I can just load everything into the state attribs assignments.

balloob commented on homeassistant/components/switch/wemo.py:
```diff
@@ -109,23 +120,88 @@ def device_state_attributes(self):
     @property
-    def current_power_mwh(self):
-        """Current power usage in mWh."""
+    def power_total_mw_min(self):
```
You can't just rename overrides of official properties. Please check out the switch ABC and see what is supported and what is not.

jumpkick replied:
The Insight returns average milliwatt power consumed for each minute.  So for 12W it will increment this value by 12,000 each minute.  So it's mW/min.

mWh is average milliwatts per hour...  The Insight doesn't provide that.

pavoni replied:
At some point I hope to refactor some of the HA power/energy code - but this requires changing some of the base components properties - and any other components that override then.

Until then I think you need to leave the official properties as they are - and add`device_state_properties` for the correct / new attributes.

pavoni commented:
I've released your `pywemo` changes as `0.4.13` ready for when you have time to wrap up this PR.

lwis commented on homeassistant/components/switch/wemo.py:
```diff
@@ -20,6 +21,16 @@
 
+# Wemo Insight
+ATTR_POWER_CURRENT_W = 'power_current_w'
+# ATTR_POWER_AVG_W = 'power_average_w'
```
Is this required?

pavoni commented:
@jumpkick Let me know if you need some help on this.

My suggestion would be to leave the overridden switch attributes (`current_power_mwh` and `today_power_mw`) as they are for this PR. 

Some of the issues you've found with these properties should be reworked as a broader PR to change all these attributes across all devices (https://github.com/home-assistant/home-assistant/issues/3828)).

The other attributes you want to include can just be added as device_state_attributes.

jumpkick commented:
@pavoni yeah, I'm thinking of starting over on this patch...  leave the stuff that's there and just load up device_state_attributes w/ the additional attributes.  I agree with #3828, but not sure how to incorporate it.

deftdawg committed:
Update device_state_attributes only

This gets rid of the other stuff and just updates device_state_attributes, leaving the default properties alone.

pavoni commented on homeassistant/components/switch/wemo.py:
```diff
@@ -109,12 +110,33 @@ def device_state_attributes(self):
 
         if self.insight_params or (self.coffeemaker_mode is not None):
             attr[ATTR_CURRENT_STATE_DETAIL] = self.detail_state
+            attr['current_power_w'] = \
```
Don't expect  the coffee machine will support these new attributes - so I'd change this so that you only retrieve these attributes for the insight device.

pavoni replied:
This is just the existing switch attribute divided by 1000.

At some point we'll change the switch attribute to be `W` rather than `mW` anyway.  So this is only really helpful until we do that.

My suggestion would be to remove it

jumpkick replied:
Good call

pavoni commented on homeassistant/components/switch/wemo.py:
```diff
@@ -109,12 +110,33 @@ def device_state_attributes(self):
+            attr['current_power_w'] = \
+                self.insight_params['currentpower'] / 1000
+            attr['today_power_mW_min'] = self.insight_params['todaymw']
+            attr['total_power_mW_min'] = self.insight_params['totalmw']
```
This line is duplicated - but it's also a duplicate of the existing `current_power_mwh` property of the switch - so you can just remove these - and depend on the switch attributes.

jumpkick replied:
Will remove in favour of the incorrectly labeled existing ones

deftdawg committed:
Incorporate comment suggestions

- Separate attribs from coffeemaker condition
- Set power units for threshold to mW to be consistent with others
- Adjust on-time labels to be more clear

pavoni commented:
👍  Looks good to me. 

@balloob are you happy for me to merge?

lwis commented on homeassistant/components/switch/wemo.py:
```diff
@@ -110,11 +111,30 @@ def device_state_attributes(self):
             attr[ATTR_CURRENT_STATE_DETAIL] = self.detail_state
 
+        if self.insight_params:
+            attr['on_latest_time'] = \
```
Can we move these to ATTR_* constants as above?

jumpkick replied:
We could put them in ATTR_ constants, but they are probably all going to be renamed and content replaced as part of addressing #3828... more places to change, so should I do it anyway?

pavoni replied:
The only one you'd need to rename is the threshold in `mw`.

Perhaps switch it to `W’ and break out the attributes in constants. Then nothing will need to be changed if my PR gets merged.

balloob commented on homeassistant/components/switch/wemo.py:
```diff
@@ -110,11 +111,30 @@ def device_state_attributes(self):
+    def as_uptime(_seconds):
+        """Format seconds into uptime string in the format: 00d 00h 00m 00s."""
+        uptime = datetime(1, 1, 1) + timedelta(seconds=_seconds)
+        return "{:0>2d}d {:0>2d}h {:0>2d}m {:0>2d}s".format(uptime.day-1,
```
Why not use the timedelta object directly? It supports days, hours etc https://docs.python.org/3/library/datetime.html#datetime.timedelta

jumpkick replied:
I'm probably doing something stupid (python noob), but when I try that I get this:
```
>>> from datetime import datetime, timedelta
>>> uptime = timedelta(days=0,seconds=86401,hours=0,minutes=0)
>>> print ("{:0>2d}d {:0>2d}h {:0>2d}m {:0>2d}s".format(uptime.days, uptime.hours, uptime.minutes, uptime.seconds))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'datetime.timedelta' object has no attribute 'hours'
```

balloob replied:
Oh you're right, I was wrong. Looks like timedelta converts it all to seconds.

robbiet480 commented:
Thanks! 🐬 🍪 💯
